include/util_quda.h

```diff
@@ -7,15 +7,42 @@
 #include <tune_key.h>
 #include <malloc_quda.h>
 
+#ifndef QUDA_MUSTTAIL
+#ifdef __clang__
+#define QUDA_MUSTTAIL __attribute__((musttail))
+#else
+#define QUDA_MUSTTAIL
+#endif
+#endif
+
 namespace quda
 {
   // strip path from __FILE__
   constexpr const char *str_end(const char *str) { return *str ? str_end(str + 1) : str; }
   constexpr bool str_slant(const char *str) { return *str == '/' ? true : (*str ? str_slant(str + 1) : false); }
   constexpr const char *r_slant(const char *str) { return *str == '/' ? (str + 1) : r_slant(str - 1); }
   constexpr const char *file_name(const char *str) { return str_slant(str) ? r_slant(str_end(str)) : str; }
+
+  template<int A, int B, int D=1, typename F>
+  __attribute__((always_inline)) __host__ __device__ inline void static_for(F&&f)
```

Comment:
Maybe this function should be target dependent?  On CUDA for example, there's no reason why this shouldn't just correspond to an unroll loop?

Reply:
It's a general utility that does not really care where it's used.

Reply:
My point is that we may want the freedom to change the `static_for` implementation for different targets.  For CUDA, we may want this to correspond to a simple unrolled for loop, and with OMP we may want it to be what you have above.

Reply:
Makes sense.  I'm not sure how to do it without introducing more macros, though.